README.md
# Linux-Shell


Basic Description :
In the shell, first I take user input form the user and then I split that string input into tokens. Then, if the given command is an external command, then, first
I create a child process (copy of parent parent process) using fork system call and in that child process I create a new process, which replaces the child process 
using the execv() system call and performs the desired work and then terminates (using exit system call). After that process terminates, then we come again back to 
the shell and then user can give input.


Commands, their options and their error handling :

1) ls : Prints the files and folders in working directory.

Options :
I. -F : Prints the files and fodders in working directory. The folder name will be ended by “/”
II. --help : Prints information about this command.

Errors handled :
I. If any other option other than the above two options is provided, then it will give error message.
II. If arguments (or options) greater than are passed, then it will give error message.
III. If the there is error in the spelling of ls entered by user, then it will give error.
IV. If the system call opendir() fails, due to some system failure, then it will give error message.

Test cases:
$ ls
$ ls -F
$ ls --help

2) cat:
If you do not provide any argument and just press enter, then whatever text you will print, it will print exactly the same text again.
If you specify the file names as arguments, then it will print the contents of the files. (You can provide more than one file at a time as argument).

Options:
a. -n: It will print the contents of the files by printing the line number also.
b. --help: Prints information about this command.

Errors handled:
I. If you just write the option -n and do not provide any file name as an argument, it will
print error message/
II. If you specify a file name which does not exist, then it will print error message.

Test case:
$ cat
Hello (You have to type this also or any text you like)
$ cat testFile.txt (A file name testFile.txt is already present in the folder Assign2)
$ cat -n testFile.txt
$ cat --help

3) date: Prints the date and time in the default time zone of linux ubuntu system. (Generally, the
default time zone of a linux system is PDT(Pacific Day Light time), so it prints according to that. )

Options:
I. -u:
It prints the date and time according to GMT time zone.
II. --help: Prints information about this command.

Errors handled;
a. If you provide an incorrect option, then it will give error.
b. If you provide more than the required arguments and option in the date command, then it
will give error.

Test case :
$ date
$ date -u
$ date --help

4) mkdir: Creates directory.

Options:
a. -v: It prints the name of the directories, after creating it.
b. --help: Prints information about this command.

Errors handled:
a. If you do not provide correct options, then it will give error.
b. If you misspell the spell of mkdir, then it will give error.

Test case:
$ mkdir testDir2
$ mkdir -v testDir3
$ mkdir --help

5) rm: Removes the specified files.

Options:
a. -i: It asks a warning message that do you really want to delete the file. (Answer has to
entered either y or n).
b. --help: Prints information about this command.

Errors handled:
a. If you use -i option and in the warning message, you entered other that y or n, then it will
give error and ask you again to enter.
b. If you specify file name to be deleted, which does not exist, then it will give error.

Test case:
$ rm testFile2.txt (testFile2.txt has been provided)
$ rm -i testFile3.txt (testFile3.txt has been provided)
$ rm --help

6) cd: Changes the directory. (note that you can specify a directory whose name has spaces, then
enquote in double quotes).

Options:
a. -L: If you use this option, then you can provide the symbolic path.
b. --help: Prints information about this command.

Errors handled:
a. If you specify a directory path which does not exist, then it will give error.
b. If you provide incorrect options, then it will give error.

Test case:
Please first create a directory, testDir.
$ cd testDir
$ cd -L testDir
$ cd --help

7) echo: Prints whatever you have written after it.

Options:
a. -n: it print whatever you have supplied and then it prints the “terminal line tag” in the
same line.
b. --help: Prints information about this command.

Errors/corner case handled:
a. There is a corner case in which the user just provide the option -n but does not write
after that what has to printed, then, then in that case, no error will be thrown and
terminal tag line will be printed in the next line.
b. If you don’ t provide the correct options, then it will give error.
c. If you mis-spell echo, then it will give error.

Testcase:
$ echo hello
$ echo -n hello
$ echo --help

8) pwd: Prints the working directory.

Options:
1) -L : Prints the symbolic path.
2) --help: : Prints information about this command.

Errors:
a. If you provide incorrect options, then it will give error.
b. If you provide arguments or options more than the required, then it will give error.

Test case:
$ pwd
$ pwd -L
$ pwd --help

9) history: Prints the history of the commands. (Note that in the output console, that history
command though which you just entered into the history console, in that console that recent
history command will not be included.)

Options:
a. -a: It will append the commands from the history list to the history.txt file, where all
commands are stored.
b. --help: Prints information about this command.

Errors:
a. If you provide incorrect options, then it will give error.
b. If you supply more arguments or options than required , it will print error.
c. If you mis-spell history, then it will give error.

Testcase:
$ history
$ history -a
$ history --help

10) exit: Exits the shell.

Options:
a. --version : Prints the version information.
b. --help: Prints information about this command.

Errors:
a. If you provide incorrect options, then it will give error.
b. If you supply more arguments or options than required , it will print error.
c. If you mis-spell exit, then it will give error.

Test case :
$ exit
$ exit --version
$ exit -- help


history.txt

1  ls
2  echo cool
3  history


lsFile.c
#include <stdio.h> 
#include <dirent.h> 
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include<stdlib.h>
#include<string.h>
  
int main(int argc, char const *argv[]) 
{ 
    if (argc == 1) {
        struct dirent *dirEnt; 
        DIR *dir = opendir("."); 
        if (dir == NULL) {
            printf("Error : Current working directory can't be opened"); 
            exit(1);
        } 
        while ((dirEnt = readdir(dir)) != NULL) {
            if ((strcmp(dirEnt->d_name, ".") != 0) && (strcmp(dirEnt->d_name, "..") != 0)) {              
                printf("%s\n", dirEnt->d_name);
            }   
        }
        closedir(dir);     
        exit(0);
    }
    else if (argc == 2 && (strcmp(argv[1], "-F") == 0)) {
        struct dirent *dirEnt; 
        DIR *dir = opendir("."); 
        if (dir == NULL) {
            printf("Error : Current working directory can't be opened"); 
            exit(1);
        } 
        while ((dirEnt = readdir(dir)) != NULL) {
            if ((strcmp(dirEnt->d_name, ".") != 0) && (strcmp(dirEnt->d_name, "..") != 0)) {
                if (dirEnt->d_type == DT_DIR) {
                    printf("%s/\n", dirEnt->d_name);
                }
                else {
                    printf("%s\n", dirEnt->d_name);
                }
            }   
        }
        closedir(dir);     
        exit(0);
    }
    else if (argc == 2 && (strcmp(argv[1], "--help") == 0)) {
        printf("ls :  Prints the files and folders in working directory.\n");
        printf("Options:\n");
        printf("-F : Prints the files and fodders in working directory. The folder name will be ended by forward slash\n");
        printf("–help : Prints information about this command\n");
        exit(0);
    }
    else {
        printf("Error: There is an error in the arguments or options.\n");
        exit(1);
    }
} 

rmFile.c
#include<stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include<stdlib.h>
#include<string.h>

int main(int argc, char const *argv[])
{	
	if ((argc >= 3) && (strcmp(argv[1], "-i") == 0)) {
		for (int i = 2; i < argc; i++) {
			printf("rm: Do you really want to delete the file %s (y/n)", argv[i]);
			char ans[100];
			scanf("%s", ans);
			//printf("%s", ans);
			//printf("cool");
			if (strcmp(ans, "y") == 0) {
				int status = remove(argv[i]);
				if (status == -1) {
					printf("Error: Directory %s can't be created\n", argv[i]);
				}
			}
			else if (strcmp(ans, "n") == 0) {

			}
			else {
				printf("You can only give y or n as input\n");
				i = i - 1;
			}
		}
		exit(0);
	} 
	else if ((argc == 2) && (strcmp(argv[1], "--help") == 0)) {
		printf("rm: Removes the specified files\n");
		printf("Options:\n");
		printf("-i: It asks a warning message that do you really want to delete the file. (Answer has to entered either y or n).\n");
		printf("–help: Prints information about this command.\n");
		exit(0);
	}
	else if ((argc >= 2) && (strcmp(argv[1], "-i") != 0)) {
		for (int i = 1; i < argc; i++) {
			int status = remove(argv[i]);
			if (status == -1) {
				printf("Error: File %s can't be deleted\n", argv[i]);
			}
		}
		exit(0);
	}

	
	
	else {
		printf("Error : There is an error in the arguments or options\n");
		exit(1);
	}

	/*
	int status = remove("test1.txt");
	printf("%d", status);
	*/
}

catFile.c
#include<stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include<stdlib.h>
#include<string.h>

int main(int argc, char const *argv[])
{
	if (argc == 1) {
		while (1) {
			char* line;
			size_t bSize = 0;
			int status = getline(&line, &bSize, stdin);          // lines are having new line character at the end
			printf("%s", line);
			if (status == -1) {
				if (feof(stdin)) {
      				exit(0);  
    			}
    			else {
    				printf("Error: There occurs a system call error while reading the input\n.");
    				exit(1);
    			} 
			}
		}
	}
	else if ((argc == 2) && (strcmp(argv[1], "--help") == 0)) {
		printf("If you do not provide any argument and just press enter, then whatever text you will print, it will print exactly the same text again\n");
		printf("If you specify the file names as arguments, then it will print the contents of the files. (You can provide more than one file at a time as argument).\n");
		printf("Options: \n");
		printf("-n: It will print the contents of the files by printing the line number also.\n");
		printf("2)	–help: Prints information about this command\n");
		exit(0);
	}
	else if ((argc >= 2) && (strcmp(argv[1], "-n") != 0)) {                     
		for (int i = 1; i < argc; i++) {
			FILE *fptr = fopen(argv[i], "r");
			if (fptr == NULL) {          
				printf("Error: File %s does not exist", argv[i]);
				exit(1);                  // Tujhe isko use karne ke liye niche waitpid() mein parameter 0 se hatkar kuch aur karna padega.
			}
			char line[350];
			//int lineNum = 1;
			while (fgets(line, sizeof(line), fptr) != NULL) {
				printf("%s", line);
			}
			fclose(fptr);
		}
		exit(0);
	}
	else if ((argc >= 3) && (strcmp(argv[1], "-n") == 0)) {
		int lineNum = 1;
		int isLineNumInc = 1;
		for (int i = 2; i < argc; i++) {
			FILE *fptr = fopen(argv[i], "r");
			if (fptr == NULL) {          
				printf("Error: File %s does not exist", argv[i]);
				exit(1);                  // Tujhe isko use karne ke liye niche waitpid() mein parameter 0 se hatkar kuch aur karna padega.
			}
			char line[350];
			while (fgets(line, sizeof(line), fptr) != NULL) {
				if (isLineNumInc == 1) {
					printf("    %d  %s", lineNum, line);
				}
				else {
					printf("%s", line);
				}
				
				int len = strlen(line);
				if (line[len - 1] == '\n') {
					lineNum = lineNum + 1;
					isLineNumInc = 1;
				}
				else {
					isLineNumInc = 0;
				}
			}
			fclose(fptr);
		}
		exit(0);
	}
	else {
		printf("Error: Error : There is an error in the arguments or options\n");
		exit(1);
	}
}

dateFile.c
#include<stdio.h>
#include<time.h>
#include<string.h>
#include<stdlib.h>

void printGMT();
char* month(int monthNo);
char* weekDay(int dayNo);
void printTime();
void help();

int main(int argc, char const *argv[])
{
	//printf("Entered");
	if (argc == 1) {
		printTime();
		exit(0);
	}
	else if (argc == 2 && (strcmp(argv[1], "-u") == 0)) {
		printGMT();
		exit(0);
	}
	else if (argc == 2 && (strcmp(argv[1], "--help") == 0)) {
		help();
		exit(0);
	}
	else {
		printf("Error: There is an error in the arguments or options.\n");
		exit(1);
	}
    
    
    /*
    printf("%d\n", argc);            // Tu no of arguments argc se check kar yeh Last NULL ko count nahi karta hain.
    printf("%s\n", argv[0]);
    printf("%s\n", argv[1]);
    printf("%s\n", argv[2]);
    */
    //printf("%s", month(1));
	//return 0;
}

/*
int main()
{
}
*/

void printTime() {
	time_t curTime;   
    time(&curTime);
    printf("%s", ctime(&curTime));
}


void printGMT() {
	time_t curTime;   // not a primitive datatype
    time(&curTime);
	struct tm *GMtime = gmtime(&curTime);
	printf ("%s %s %d %2d:%02d:%02d %d\n", weekDay(GMtime->tm_wday), month(GMtime->tm_mon), GMtime->tm_mday, GMtime->tm_hour % 24, GMtime->tm_min, GMtime->tm_sec, 1900 + GMtime->tm_year);       // %02d ka work dekh

}

char* month(int monthNo) {
	char* arr[] = {"Jan", "Feb", "Mar", "Apr", "May", "June", "July", "Aug", "Sep", "Oct", "Nov", "Dec"}; 
	return arr[monthNo];
}

char* weekDay(int dayNo) {
	char* arr[] = {"Sun", "Mon", "Tue", "Wed", "Thurs", "Fri", "Sat"};
	return arr[dayNo];
}

void help() {
	printf("Prints the date and time in the default time zone of linux ubuntu system. (Generally, the default time zone of a linux system is PDT(Pacific Day Light time), so it prints according to that. )\n"); 
	printf("Options:\n");
	printf("-u: It prints the date and time according to GMT time zone\n");
	printf("–help: Prints information about this command\n");
}







mkdirFile.c
#include<stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include<stdlib.h>
#include<string.h>


// Tu shell.c file mein /name/ ke andar wale strings ko ek saath karke yahan bhej
//  TU special charater /name/ ka use kar for above purpose 


int main(int argc, char const *argv[])
{
	//printf("Entered\n");
	if ((argc == 2) && (strcmp(argv[1], "--help") == 0)) {
		printf("Creates Directory\n");
		printf("Options:\n");
		printf("-v: It prints the name of the directories, after creating it.\n");
		printf("--help: Prints information about this command\n");
		exit(0);
	}
	else if ((argc >= 3) && (strcmp(argv[1], "-v") == 0)) {
		for (int i = 2; i < argc; i++) {
			int status = mkdir(argv[i], 0777);
			if (status == -1) {
				printf("Error: Directory %s can't be created\n", argv[i]);
			}
			else {
				printf("mkdir: created Directory %s\n", argv[i]);
			}
		}
		exit(0);
	} 
	else if (argc >= 2 && (argv[1][0] == '"' && argv[argc - 1][strlen(argv[argc - 1]) - 1] == '"')) {
		char myDir[400] = "";
		//argv[argc - 1][strlen(argv[argc - 1]) - 1] = '\0';
		//argv[1] = argv[1] + 1;
		for (int j = 1; j < argc; j++) {
			strcat(myDir, argv[j]);
			if (j != argc - 1) {
				strcat(myDir, " ");
			}
		}
		//myDir = myDir + 1;
		int len = strlen(myDir);
		myDir[len-1] = '\0';
		char* myDir1 = myDir;
		myDir1 = myDir1 + 1;
		mkdir(myDir1, 0777);
		exit(0);
	}
	else if ((argc >= 2) && (strcmp(argv[1], "-v") != 0)) {
		for (int i = 1; i < argc; i++) {
			int status = mkdir(argv[i], 0777);
			if (status == -1) {
				printf("Error: Directory %s can't be created\n", argv[i]);
			}
		}
		exit(0);
	}

	
	
	else {
		printf("Error : There is an error in the arguments or options\n");
		exit(1);
	}
	
	/*int status = mkdir("testDir1/testDir2", 0777);       // Agar dirctory pehle se hi hain, toh bhi yeh bana deta hain. So, done
	if (status == -1) {
		printf("Error: Directory can't be created.");
		exit(1);
	}
	else {
		exit(0);
	}*/
}



shell.c
#include<stdio.h>
#include <unistd.h> 
#include<sys/wait.h>
#include<stdlib.h> 
#include<string.h>
#include <limits.h>


void printWorkingDir();


int main() {
	char* intComArr[] = {"cd", "echo", "history", "pwd", "exit"};
	char* extComArr[] = {"date", "ls", "cat", "rm", "mkdir"};
	char* baseAddress = "/home/akhil/Desktop/Assign2/";
	char* extComFileArr[] = {"dateFile", "lsFile", "catFile", "rmFile", "mkdirFile"};
	
	
	while (1) {
		char cwd[PATH_MAX];
		getcwd(cwd, PATH_MAX);
		printf("akhil@shell:%s $ ", cwd);      

		char* line;
		size_t bSize = 0;
		getline(&line, &bSize, stdin);          // lines are having new line character at the end
		

		
		FILE *fptr1 = fopen("historyNum.txt", "r");
		int hisNoAvailable;
		fscanf(fptr1, "%d", &hisNoAvailable);
		
		fptr1 = fopen("historyNum.txt", "w");
		fprintf(fptr1, "%d", hisNoAvailable + 1);
		//char arr[] = "hello";
		//char str[] = "hello";
		//char arr[] = "hello";
		
		FILE *fptr2 = fopen("historyList.txt", "a");
		//char num[100];
		//sprintf(num, "%d", hisNoAvailable);
		fprintf(fptr2, "%d  %s", hisNoAvailable, line);
		



		//fclose(fptr1);
		//fptr1 = fopen("historyNum.txt", "r");
		//fopen("historyNum.txt", "a");
		//fopen("historyList.txt", "r");
		//fclose(fptr1);
		//fclose(fptr2);
		//free(fptr1);
		//free(fptr2);


		int len = strlen(line);
		line[len - 1] = '\0';                   // removed the new line character. String length also charges
		int noOfTokens = 1;
		for (int i = 0; i < strlen(line); i++) {
			if (line[i] == ' ') {
				noOfTokens = noOfTokens + 1;
			}
		}
		noOfTokens = noOfTokens + 1;
		//printf("%d\n", noOfTokens);
		//noOfTokens = noOfTokens + 1;  // for NULL
		


		char* tokens[noOfTokens];
		char* token = strtok(line, " ");
		int ind = 0;
		while (token != NULL) {
			tokens[ind] = token;
			token = strtok(NULL, " ");
			ind = ind + 1;
		}
		tokens[ind] = NULL;


		//split(line, noOfTokens, &tokens);
		//printf("%s", tokens[0]);
		//printf("%s", tokens[2]);
		//printf("%s", tokens[2]);
		//break;
		//free(line);
		int found = 0;
		for (int i = 0; i < 5; i++) {
			if (strcmp(intComArr[i], tokens[0]) == 0) {
				found = 1;
				if (i == 0) {
					int noOfArg = noOfTokens - 1;
					if (noOfArg == 1) {
						chdir(getenv("HOME"));
					}
					else if (noOfArg == 2 && (strcmp(tokens[1], "~") == 0)) {
						chdir(getenv("HOME"));
					}
					else if (noOfArg == 2 && (strcmp(tokens[1], "-L") == 0)) {
						chdir(getenv("HOME"));
					}
					else if (noOfArg == 2 && (strcmp(tokens[1], "--help") == 0)) {
						printf("Changes the directory. (note that you can specify a directory whose name has spaces, then enquote in double quotes).\n");
						printf("Options:\n");
						printf("-L: If you use this option, then you can provide the symbolic path\n");
						printf("–help: Prints information about this command.\n");
						
					}
					else if (noOfArg == 3 && (strcmp(tokens[1], "-L") == 0)) {
						int status = chdir(tokens[2]);
						if (status != 0) {
							printf("Error: The directory does not exist\n");
						}
					}
					else if (noOfArg >= 2 && (tokens[1][0] == '"' && tokens[noOfArg - 1][strlen(tokens[noOfArg - 1]) - 1] == '"')) {
						char myDir[400] = "";
						tokens[noOfArg - 1][strlen(tokens[noOfArg - 1]) - 1] = '\0';
						tokens[1] = tokens[1] + 1;
						for (int j = 1; j < noOfArg; j++) {
							strcat(myDir, tokens[j]);
							if (j != noOfArg - 1) {
								strcat(myDir, " ");
							}
						}
						int status = chdir(myDir);
						if (status != 0) {
							printf("Error: The directory does not exist\n");
						}
					}
					else if (noOfArg == 2) {
						int status = chdir(tokens[1]);
						if (status != 0) {
							printf("Error: The directory does not exist\n");
						}
					}
					else {
						printf("Error: There is an error in the arguments or options.\n");
					}
				}
				else if (i == 1) {
					int noOfArg = noOfTokens - 1;
					if (noOfArg == 1) {
						printf("\n");
					}
					else if ((noOfArg == 2) && (strcmp(tokens[1], "-n") == 0)) {
						//printf("");
					}
					else if ((noOfArg == 2) && (strcmp(tokens[1], "--help") == 0)){
						printf("Prints whatever you have written after it.\n");
						printf("Options:\n");
						printf("-n: it print whatever you have supplied and then it prints the “terminal line tag” in the same line.\n");
						printf("–help: Prints information about this command.\n");
					}
					else if ((noOfArg >= 2) && (strcmp(tokens[1], "-n") != 0)) {
						for (int j = 1; j < noOfArg; j++) {
							for (int k = 0; k < strlen(tokens[j]); k++) {
								if (tokens[j][k] == '\n') {
									printf("n");
								}
								else if (tokens[j][k] == '\t') {
									printf("t");
								}
								else if (tokens[j][k] == '\\') {
									printf("\\");
								}
								else if (tokens[j][k] == '\b') {
									printf("b");
								}
								else if (tokens[j][k] == '\a') {
									printf("a");
								}
								else if (tokens[j][k] == '\v') {
									printf("v");
								}
								else if (tokens[j][k] == '\0') {
									printf("0");
								}
								else if ((tokens[j][k] == '"') && (k == 0 || k == strlen(tokens[j]) - 1) && ((tokens[j][0] == '"') && (tokens[j][strlen(tokens[j]) - 1] == '"'))) {

								}
								else {
									printf("%c", tokens[j][k]);
								}

							}
							if (j == noOfArg - 1) {
								printf("\n");
							}
							else {
								printf(" ");
							}
						}
					}
					else if ((noOfArg >= 3) && (strcmp(tokens[1], "-n") == 0)) {
						for (int j = 2; j < noOfArg; j++) {
							for (int k = 0; k < strlen(tokens[j]); k++) {
								if (tokens[j][k] == '\n') {
									printf("n");
								}
								else if (tokens[j][k] == '\t') {
									printf("t");
								}
								else if (tokens[j][k] == '\\') {
									printf("\\");
								}
								else if (tokens[j][k] == '\b') {
									printf("b");
								}
								else if (tokens[j][k] == '\a') {
									printf("a");
								}
								else if (tokens[j][k] == '\v') {
									printf("v");
								}
								else if (tokens[j][k] == '\0') {
									printf("0");
								}
								else if ((tokens[j][k] == '"') && (k == 0 || k == strlen(tokens[j]) - 1) && ((tokens[j][0] == '"') && (tokens[j][strlen(tokens[j]) - 1] == '"'))) {
									
								}
								else {
									printf("%c", tokens[j][k]);
								}

							}
							if (j == noOfArg - 1) {
								//printf("");
							}
							else {
								printf(" ");
							}
						}
					}
					else {
						printf("Error : There is an error in the arguments or options\n");
					}
				}
				
				else if (i == 2) {
					int noOfArg = noOfTokens - 1;
					if (noOfArg == 1) {
						FILE *fptr3 = fopen("history.txt", "r");
						char line1[350];
						//int lineNum = 0;
						while (fgets(line1, sizeof(line1), fptr3) != NULL) {
							printf("%s", line1);
						}
						fclose(fptr3);

						FILE *fptr4 = fopen("historyList.txt", "r");
						//char line2[350];
						//int lineNum = 0;
						while (fgets(line1, sizeof(line1), fptr4) != NULL) {
							printf("%s", line1);
						}
						fclose(fptr4);
						//free(fptr3);
						//free(fptr4);
						//free(line1);
					}
					else if (noOfArg == 2 && (strcmp(tokens[1], "-a") == 0)) {
						FILE *fptr3 = fopen("historyList.txt", "r");
						FILE *fptr4 = fopen("history.txt", "a");
						char line1[350];
						while (fgets(line1, sizeof(line1), fptr3) != NULL) {
							fprintf(fptr4, "%s", line1);
						}
						fclose(fptr3);
						fclose(fptr4);
						fptr3 = fopen("historyList.txt", "w");
						fclose(fptr3);
						//free(fptr3);
						//free(fptr4);
						//free(fptr5);
						//free(line1);
					}
					else if (noOfArg == 2 && (strcmp(tokens[1], "--help") == 0)) {
						printf("history: prints history.\n");
					}
					else {
						printf("Error: There is an error in the arguments or options.\n");
					}
				}
				else if (i == 3) {
					int noOfArg = noOfTokens - 1;
					if (noOfArg == 1) {
						printWorkingDir();
					}
					else if (noOfArg == 2 && (strcmp(tokens[1], "-L")== 0)) {
						printWorkingDir();
					}
					else if (noOfArg == 2 && (strcmp(tokens[1], "--help") == 0)) {
						printf("prints working directory\n");
					}
					else {
						printf("Error : There is an error in the arguments or options\n");
					}
					
				}
				else if (i == 4) {
					int noOfArg = noOfTokens - 1;
					if (noOfArg == 1) {
						
						FILE *fptr3 = fopen("historyList.txt", "r");
						FILE *fptr4 = fopen("history.txt", "a");
						char line1[350];
						while (fgets(line1, sizeof(line1), fptr3) != NULL) {
							fprintf(fptr4, "%s", line1);
						}
						fclose(fptr3);
						fclose(fptr4);
						fptr3 = fopen("historyList.txt", "w");
						fclose(fptr3);
						//free(fptr3);
						//free(fptr4);
						//free(fptr5);
						//free(line1);
						
						exit(0);
					}
					else if (noOfArg == 2 && (strcmp(tokens[1], "--version") == 0)) {
						printf("mkdir: Version - 8.2.3\n");
						printf("Developed by Akhil Anand\n");
					}
					else if (noOfArg == 2 && (strcmp(tokens[1], "--help") == 0)) {
						printf("exits the shell\n");
					}
					else {
						printf("Error: There is an error in the arguments or options.\n");
					}
				}

				break;
			}
		}
		if (found == 0) {
			//printf("Enter1");
			for (int i = 0; i < 5; i++) {
				if (strcmp(extComArr[i], tokens[0]) == 0) {
					//printf("Enter2\n");
					found = 1;
					int pid = fork();
					if (pid == 0) {                 // pid < 0 ka bhi kar
						//printf("child\n");
						//printf("%s\n", baseAddress);
						char* path = (char *) malloc(1 + strlen(baseAddress)+ strlen(extComFileArr[i]));
						strcat(path, baseAddress);
						strcat(path, extComFileArr[i]);
						//char* path = "";
						//strcat(path, baseAddress);
						//printf("%s\n", path);
						//printf("%s\n", baseAddress);
						//strcat(path, extComFileArr[i]);
						//printf("%s\n", path);
						tokens[0] = path; 
						//char* args[]={"/home/akhil/Desktop/csvAssign/dateFile", "-u", NULL}; 
        				execv(tokens[0], tokens); 
					}
					else {
						wait(NULL);
					}
					break;
				}
			}
		}

		if (found == 0) {
			printf("Error : The command is not recognised.\n");
		}
		fclose(fptr1);
		fclose(fptr2);
		
		
	}
	exit(0);
}


void printWorkingDir() {
	char cwd[PATH_MAX];
	if (getcwd(cwd, PATH_MAX) != NULL) {
       printf("%s\n", cwd);
   	} 
   	else {
       printf("Error : Working directory can't be printed\n");
   	}
}







